iceoryx_utils/test/moduletests/test_cxx_forward_list.cpp

```diff
@@ -0,0 +1,1931 @@
+
+#include "iceoryx_utils/cxx/forward_list.hpp"
+#include "test.hpp"
+
+
+using namespace ::testing;
+using namespace iox::cxx;
+
+static constexpr uint64_t TESTLISTCAPACITY{10u};
+static constexpr int C_TOR_TEST_VALUE_DEFAULT_VALUE{-99};
+
+class forward_list_test : public Test
+{
+  public:
+    static int cTor;
+    static int customCTor;
+    static int copyCTor;
+    static int moveCTor;
+    static int moveAssignment;
+    static int copyAssignment;
+    static int dTor;
+    static int classValue;
+
+    class CTorTest
+    {
+      public:
+        CTorTest()
+        {
+            cTor++;
+            classValue = m_value;
+        }
+
+        CTorTest(const int value)
+            : m_value(value)
+        {
+            customCTor++;
+            classValue = m_value;
+        }
+
+        CTorTest(const CTorTest& rhs)
+        {
+            copyCTor++;
+            m_value = rhs.m_value;
+            classValue = m_value;
+        }
+
+        CTorTest(CTorTest&& rhs)
+        {
+            moveCTor++;
+            m_value = rhs.m_value;
+            classValue = m_value;
+        }
+
+        CTorTest& operator=(const CTorTest& rhs)
+        {
+            copyAssignment++;
+            m_value = rhs.m_value;
+            classValue = m_value;
+            return *this;
+        }
+
+        CTorTest& operator=(CTorTest&& rhs)
+        {
+            moveAssignment++;
+            m_value = rhs.m_value;
+            classValue = m_value;
+            return *this;
+        }
+
+        bool operator==(const CTorTest& rhs) const
+        {
+            return rhs.m_value == m_value;
+        }
+
+        ~CTorTest()
+        {
+            dTor++;
+            classValue = m_value;
+        }
+
+        int m_value = C_TOR_TEST_VALUE_DEFAULT_VALUE;
+    };
+
+
+    void SetUp()
+    {
+        cTor = 0;
+        customCTor = 0;
+        copyCTor = 0;
+        moveCTor = 0;
+        moveAssignment = 0;
+        copyAssignment = 0;
+        dTor = 0;
+        classValue = 0;
+    }
+
+    bool isSetupState()
+    {
+        if (cTor == 0 && customCTor == 0 && copyCTor == 0 && moveCTor == 0 && moveAssignment == 0 && copyAssignment == 0
+            && dTor == 0 && classValue == 0)
+        {
+            return true;
+        }
+        else
+        {
+            return false;
+        }
+    }
+
+    forward_list<CTorTest, TESTLISTCAPACITY> sut;
+};
+
+// forward_list_test statics
+int forward_list_test::cTor;
+int forward_list_test::customCTor;
+int forward_list_test::copyCTor;
+int forward_list_test::moveCTor;
+int forward_list_test::moveAssignment;
+int forward_list_test::copyAssignment;
+int forward_list_test::dTor;
+int forward_list_test::classValue;
+
+namespace
+{
+// test function for iterator_traits testing
+template <typename IterType>
+int iteratorTraitReturnDoubleValue(IterType iter)
+{
+    typedef typename std::iterator_traits<IterType>::value_type IterValueType;
+    IterValueType m_value = *iter;
```

Comment:
OK, this checks that the underlying value is of the type the iterator trait indicates I guess (would not compile otherwise, or at least must be convertible).